File: py/parse.py
from typing import Callable
from dataclasses import dataclass
from enum import Enum, auto

from vm import Op
from tokenizer import Tokenizer, Token, TokenType


class ParseError:
    def __init__(self, token: Token, message: str):
        self.token = token
        self.message = message

    def __str__(self):
        if self.token.ty == TokenType.EOF:
            at_text = "at end"
        else:
            at_text = f"at '{self.token.value}'"
        return f"[line {self.token.line}] ERROR: {self.message} {at_text}"


class Parse:
    def __init__(self, source: str) -> None:
        self.tokenizer = Tokenizer(source)
        self.current = None
        self.previous = None
        self.errors = []
        self.has_error = False
        self.panic_mode = False

    def advance(self) -> Token:
        self.previous = self.current
        self.current = self.tokenizer.scan_token()
        return self.current

    def error_at_current(self, message: str) -> None:
        self.error_at(self.current, message)

    def error_at(self, token: Token, message: str) -> None:
        error = ParseError(token, message)
        if not self.panic_mode:
            print(error)

        self.errors.append(error)
        self.has_error = True
        self.panic_mode = True

File: src/Parse.zig
const Self = @This();

tokenizer: Tokenizer,
hadError: bool = false,
panicMode: bool = false,
current: Tokenizer.Token = undefined,
previous: Tokenizer.Token = undefined,

pub fn init(source: []const u8) Self {
    var tokenizer = Tokenizer.init(source);
    return .{ .tokenizer = tokenizer };
}

pub fn advance(self: *Self) void {
    self.previous = self.current;
    while (true) {
        self.current = self.tokenizer.scanToken();
        if (self.current.tag != .invalid) break;
        self.errorAtCurrent("invalid token");
    }
}

pub fn errorAtCurrent(self: *Self, message: []const u8) void {
    self.errorAt(&self.current, message);
}

pub fn errorAtPrevious(self: *Self, message: []const u8) void {
    self.errorAt(&self.previous, message);
}

pub fn errorAt(self: *Self, token: *Tokenizer.Token, message: []const u8) void {
    if (self.panicMode) return;
    self.panicMode = true;
    std.debug.print("[line {d}] Error", .{token.loc.lineno});
    if (token.tag == .eof) {
        std.debug.print(" at end.", .{});
    } else if (token.tag == .invalid) {} else {
        std.debug.print(" at '{s}'", .{self.tokenizer.buffer[token.loc.start..token.loc.end]});
    }

    std.debug.print(" {s}\n", .{message});

    self.hadError = true;
    self.panicMode = true;
}

const std = @import("std");
const Tokenizer = @import("./Tokenizer.zig");

test "Parse" {
    const Parse = @This();

    var parse = Parse.init("1 + 1");
    parse.advance();
    parse.advance();
}
